### PR TITLE
ReadMany: Adds ReadManyItems Api using Query under the covers

### PR DESCRIPTION
**Pull request by asketagarwal**
## Sample

Stream:
```
  IReadOnlyList<(string, PartitionKey)> itemList = new List<(string, PartitionKey)>
  {
       ("Id1", new PartitionKey("pkValue1")),
       ("Id2", new PartitionKey("pkValue2")),
       ("Id3", new PartitionKey("pkValue3"))
  };
   
   using (ResponseMessage responseMessage = await this.Container.ReadManyItemsStreamAsync(itemList))
   {
       using (Stream stream = response.ReadBodyAsync())
       {
           //Read or do other operations with the stream
           using (StreamReader streamReader = new StreamReader(stream))
           {
               string content = streamReader.ReadToEndAsync();
           }
       }
   }
```

Typed:

```
   public class ToDoActivity{
       public string id {get; set;}
       public string status {get; set;}
   }
  
   IReadOnlyList<(string, PartitionKey)> itemList = new List<(string, PartitionKey)>
   {
       ("Id1", new PartitionKey("pkValue1")),
       ("Id2", new PartitionKey("pkValue2")),
       ("Id3", new PartitionKey("pkValue3"))
   };
  
  FeedResponse<ToDoActivity> feedResponse = this.Container.ReadManyItemsAsync<ToDoActivity>(itemList);
```
## Closing issues

To automatically close an issue: closes #IssueNumber

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+            foreach (KeyValuePair<string, List<(string, PartitionKey)>> entry in partitionKeyRangeItemMap)
+            {
+                tasks.Add(Task.Run(async () =>
+                {
+                    // Only allow 'maxConcurrency' number of queries at a time
+                    await semaphore.WaitAsync();
+
+                    try
+                    {
+                        QueryDefinition queryDefinition = (this.partitionKeySelector == "[\"id\"]") ?
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs:52]: What about subpartitioning? PartitionKey can have multiple partition key values.

File: Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+                                                   this.CreateReadManyQueryDefifnitionForId(entry.Value) :
+                                                   this.CreateReadManyQueryDefifnitionForOther(entry.Value);
+
+                        List<ResponseMessage> pages = new List<ResponseMessage>();
+                        FeedIteratorInternal feedIterator = (FeedIteratorInternal)this.container.GetItemQueryStreamIterator(new FeedRangePartitionKeyRange(entry.Key),
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs:57]: FeedRangePartitionKeyRange will throw on splits/merge based on discussion with ealsur

File: Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+                        cosmosElements.AddRange(queryResponse.CosmosElements);
+                    }
+                    else
+                    {
+                        throw new InvalidOperationException("Read Many is being used with Query");
+                    }
+                }
+            }
+
+            return new ResponseMessage(System.Net.HttpStatusCode.OK)
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs:136]: Headers like RUs will need to be aggregated.

File: Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs (hunk 2 above)
Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs:57]: Why not use FeedRanges directly?

`container.GetFeedRangesAsync()` instead of using the PartitionKeyRanges as source?

File: Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+                        QueryDefinition queryDefinition = (this.partitionKeySelector == "[\"id\"]") ?
+                                                   this.CreateReadManyQueryDefifnitionForId(entry.Value) :
+                                                   this.CreateReadManyQueryDefifnitionForOther(entry.Value);
+
+                        List<ResponseMessage> pages = new List<ResponseMessage>();
+                        FeedIteratorInternal feedIterator = (FeedIteratorInternal)this.container.GetItemQueryStreamIterator(new FeedRangePartitionKeyRange(entry.Key),
+                                                                        queryDefinition);
+                        while (feedIterator.HasMoreResults)
+                        {
+                            using (ResponseMessage responseMessage = await feedIterator.ReadNextAsync(trace, cancellationToken))
```
Comment (abhijitpai) [inline Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs:61]: If a single page fetch call fails at a single partition, can we avoid losing the successful work done?

File: Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+                Content = CosmosElementSerializer.ToStream(string.Empty, cosmosElements, ResourceType.Document)
+            };
+        }
+
+        private QueryDefinition CreateReadManyQueryDefifnitionForId(List<(string, PartitionKey)> items)
+        {
+            StringBuilder queryStringBuilder = new StringBuilder();
+
+            queryStringBuilder.Append("SELECT * FROM c WHERE c.id IN ( ");
+            for (int i = 0; i < items.Count; i++)
```
Comment (abhijitpai) [inline Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs:147]: Consider the max query length limits that the server will enforce when you construct this.

File: Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs (hunk 4 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs:61]: The current design is to return a single page response for ease of consuming it. If none of the built in retry policy handle it I don't see any benefit in returning a partial response. If it is something that can be retried the SDK should do it automatically instead of forcing developers to handle all the different scenarios.

File: Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs (hunk 5 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs:147]: Max is 256 KB: https://docs.microsoft.com/en-us/azure/cosmos-db/concepts-limits#sql-query-limits

File: Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+        private readonly PartitionKeyDefinition partitionKeyDefinition;
+        private readonly int maxConcurrency = Environment.ProcessorCount * 10;
+        private readonly ContainerCore container;
+        private readonly CancellationToken cancellationToken;
+
+        public ReadManyQueryHelper(PartitionKeyDefinition partitionKeyDefinition,
+                                   ContainerCore container,
+                                   CancellationToken cancellationToken)
+        {
+            this.partitionKeyDefinition = partitionKeyDefinition;
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs:33]: Need to validate container recreate scenarios.

File: Microsoft.Azure.Cosmos/src/ReadManyHelper.cs [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Cosmos.Tracing;
+
+    internal abstract class ReadManyHelper
+    {
+        public abstract Task<ResponseMessage> ExecuteReadManyRequestAsync(IReadOnlyList<(string, PartitionKey)> items,
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/ReadManyHelper.cs:15]: Not why the contract differs from the mainline ReadMany contract. 

ex: 
- CancelationToken is not wired through. 
- ReqeustOptions moved to the CTOR

FWIW its better to have explicit contract and flow it through.

File: Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs [hunk 8]
```diff
@@ -0,0 +1,361 @@
...
+                                        ITrace trace)
+        {
+            SemaphoreSlim semaphore = new SemaphoreSlim(0, this.maxConcurrency);
+            List<Task<List<TResult>>> tasks = new List<Task<List<TResult>>>();
+
+            foreach (KeyValuePair<PartitionKeyRange, List<(string, PartitionKey)>> entry in partitionKeyRangeItemMap)
+            {
+                using (ITrace childTrace = trace.StartChild("Execute query for a partitionkeyrange", TraceComponent.Query, TraceLevel.Info))
+                {
+                    tasks.Add(Task.Run(async () =>
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs:83]: @ealsur will it have any impact on single scheduler environments?

File: Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs [hunk 9]
```diff
@@ -0,0 +1,361 @@
...
+                    {
+                        // Only allow 'maxConcurrency' number of queries at a time
+                        await semaphore.WaitAsync();
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs:86]: Can the guard be prior task generation?

In other words, if the task is not guarded then it might lead to pressure on task scheduler all just waiting. No?

/cc: @j82w, @FabianMeiswinkel , @ealsur

File: Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs [hunk 10]
```diff
@@ -0,0 +1,361 @@
...
+
+                        try
+                        {
+                            return await generateResponsesForPartition(entry.Value, entry.Key, childTrace);
+                        }
+                        finally
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs:92]: How are splits or merges handled?

Comment (kirankumarkolli) [reply Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs:92]: Split handling is must at-least for now.

File: Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs [hunk 11]
```diff
@@ -0,0 +1,361 @@
...
+            return new QueryDefinition(queryStringBuilder.ToString());
+        }
+
+        private QueryDefinition CreateReadManyQueryDefifnitionForOther(List<(string, PartitionKey)> items)
+        {
+            StringBuilder queryStringBuilder = new StringBuilder();
+            SqlParameterCollection sqlParameters = new SqlParameterCollection();
+
+            queryStringBuilder.Append("SELECT * FROM c WHERE ( ");
+            for (int i = 0; i < items.Count; i++)
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs:208]: Won't we need to guards on how many to pack for a single execution? 

ex: What is all 200 point to the same PKRangeId => bigger query => BE rejects?

File: Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs (hunk 10 above)
Comment (asketagarwal) [reply Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs]: we use query under the covers which itself is split proof. I am using FeedRangeEpk which will take care of split

File: Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs (hunk 11 above)
Comment (asketagarwal) [reply Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs]: max 1000 items per query

File: Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs (hunk 5 above)
Comment (asketagarwal) [reply Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs]: max 1000 items per query

File: Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs [hunk 12]
```diff
@@ -0,0 +1,403 @@
...
+                                                                this.GenerateTypedResponsesForPartitionAsync<T>,
+                                                                readManyRequestOptions,
+                                                                trace,
+                                                                cancellationToken);
+
+            return this.CombineFeedResponseFromQueryResponses(queryResponses, trace);
+        }
+
+        internal Task<List<TResult>[]> ReadManyTaskHelperAsync<TResult>(IDictionary<PartitionKeyRange, List<(string, PartitionKey)>> partitionKeyRangeItemMap,
+                                  Func<QueryDefinition, PartitionKeyRange, ReadManyRequestOptions, ITrace, CancellationToken, Task<List<TResult>>> generateResponsesForPartition,
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs:76]: Instead of passing in a function why not always just return (IReadOnlyList<CosmosElements> items, Headers aggregatedHeaders)? Then the different methods can just convert it to a stream or typed response.

File: Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs [hunk 13]
```diff
@@ -0,0 +1,403 @@
...
+            return pages;
+        }
+
+        private async Task<List<FeedResponse<T>>> GenerateTypedResponsesForPartitionAsync<T>(QueryDefinition queryDefinition,
+                                                                          PartitionKeyRange partitionKeyRange,
+                                                                          ReadManyRequestOptions readManyRequestOptions,
+                                                                          ITrace trace,
+                                                                          CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs:330]: Instead of throwing it should just return null. Throwing adds a lot of overhead and doesn't help here.

File: Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs [hunk 14]
```diff
@@ -0,0 +1,403 @@
...
+        {
+            SemaphoreSlim semaphore = new SemaphoreSlim(0, this.maxConcurrency);
+            List<Task<List<TResult>>> tasks = new List<Task<List<TResult>>>();
+
+            foreach (KeyValuePair<PartitionKeyRange, List<(string, PartitionKey)>> entry in partitionKeyRangeItemMap)
+            {
+                // Fit MaxItemsPerQuery items in a single query to BE
+                for (int startIndex = 0; startIndex < entry.Value.Count; startIndex += this.maxItemsPerQuery)
+                {
+                    using (ITrace childTrace = trace.StartChild("Execute query for a partitionkeyrange", TraceComponent.Query, TraceLevel.Info))
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs:89]: ITrace needs to be passed inside the Task.Run. This will only show the time it took to create the task. Not the time to actually do the operation. Creating the child inside the Task.Run is also an issue because it is not thread safe.

File: Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs [hunk 15]
```diff
@@ -0,0 +1,403 @@
...
+                    queryStringBuilder.Append(" OR ");
+                }
+            }
+            queryStringBuilder.Append(" )");
+
+            return QueryDefinition.CreateFromQuerySpec(new SqlQuerySpec(queryStringBuilder.ToString(),
+                                                        sqlParameters));
+        }
+
+        private List<string> CreatePkSelectors(PartitionKeyDefinition partitionKeyDefinition)
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs:270]: Why is this needed? Shouldn't the paths from the partition key definition be valid?

Comment (asketagarwal) [reply Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs:270]: I need to break down a path like "/a/b/c" into c["a"]["b"]["c"] when using it in a query.

File: Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs (hunk 12 above)
Comment (asketagarwal) [reply Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs]: avoided function passing. Using stream for both and then serializing for typed.

File: Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs (hunk 14 above)
Comment (asketagarwal) [reply Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs]: used trace.Dispose() explicitly inside task.run

File: Microsoft.Azure.Cosmos/src/Resource/Container/Container.cs [hunk 16]
```diff
@@ -620,6 +620,77 @@ public abstract class Container
...
+        ///     ("Id2", new PartitionKey("pkValue2")),
+        ///     ("Id3", new PartitionKey("pkValue3"))
+        /// };
+        ///
+        /// FeedResponse<ToDoActivity> feedResponse = this.Container.ReadManyItemsAsync<ToDoActivity>(itemList);
+        /// ]]>
+        /// </code>
+        /// </example>
+        public abstract Task<FeedResponse<T>> ReadManyItemsAsync<T>(
+            IReadOnlyList<(string, PartitionKey)> items,
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Resource/Container/Container.cs:690]: nit: named 
```suggestion
            IReadOnlyList<(string id, PartitionKey partitionKey)> items,
```

File: Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs (hunk 8 above)
Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs:90]: I have honestly no idea, but we have tests that use single scheduler, and we could try this? https://github.com/Azure/azure-cosmos-dotnet-v3/blob/master/Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/SingleTaskScheduler.cs

File: Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs (hunk 9 above)
Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs:93]: Correct, the guard should ideally be before, otherwise the Task is still being generated and just stays there waiting. If we can avoid the generation and instead hold the for loop, it might be better.

File: Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs [hunk 17]
```diff
@@ -0,0 +1,369 @@
...
+                                                                                  CancellationToken cancellationToken)
+        {
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return null;
+            }
+
+            List<ResponseMessage> pages = new List<ResponseMessage>();
+            FeedIteratorInternal feedIterator = (FeedIteratorInternal)this.container.GetItemQueryStreamIterator(
+                                                    new FeedRangeEpk(partitionKeyRange.ToRange()),
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs:318]: This is crafty, good one

**Review by ealsur — APPROVED** [review]
LGTM, the only points are around the semaphore wait that Kiran already pointed out

File: Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs (hunk 8 above)
Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs:90]: I'd argue though that this might not be a feature you'd want to run on single scheduler environments maybe

File: Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs (hunk 9 above)
Comment (asketagarwal) [reply Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs:93]: makes sense. I have added the guard before the task generation.

File: Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs [hunk 18]
```diff
@@ -0,0 +1,367 @@
...
+                                                    queryDefinition,
+                                                    continuationToken: null,
+                                                    requestOptions: readManyRequestOptions?.ConvertToQueryRequestOptions());
+            while (feedIterator.HasMoreResults)
+            {
+                try
+                {
+                    ResponseMessage responseMessage = await feedIterator.ReadNextAsync(trace, cancellationToken);
+                    responseMessage.EnsureSuccessStatusCode();
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs:325]: The Stream API of read many should not throw the exception. It should just return the ResponseMessage with the status code. Please add tests to verify it does not throw for scenarios like 404 or 429.

**Review by j82w — APPROVED** [review]

File: Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs (hunk 8 above)
Comment (kirankumarkolli) [reply Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs:91]: There are many such query aspects.

File: Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs [hunk 19]
```diff
@@ -0,0 +1,392 @@
...
+
+            foreach (KeyValuePair<PartitionKeyRange, List<(string, PartitionKey)>> entry in partitionKeyRangeItemMap)
+            {
+                // Fit MaxItemsPerQuery items in a single query to BE
+                for (int startIndex = 0; startIndex < entry.Value.Count; startIndex += this.maxItemsPerQuery)
+                {
+                    // Only allow 'maxConcurrency' number of queries at a time
+                    await semaphore.WaitAsync();
+
+                    ITrace childTrace = trace.StartChild("Execute query for a partitionkeyrange", TraceComponent.Query, TraceLevel.Info);
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs:89]: Do PKRangeId show up in the trace-data?

**Review by kirankumarkolli — APPROVED** [review]

File: Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs (hunk 19 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs:89]: Yes, query adds the PKRangeIds
